### PR TITLE
Radial effects v1

### PR DESCRIPTION
**Pull request by budadcabrion**
- radar, pulse, and spiral effects
- frequency (number of arms/pulses) and size settings
- move visualizer picker to VJ window
- uses 1 gradient

NEXT PR:
- support multiple gradients
- star effect
- better use of level and/or bass drum events
- clean up "JK TEST" stuff

### CODE REVIEW

File: Spectrum/Visualizers/LEDDomeRadialVisualizer.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    // Map value x from range a-b to range c-d
+    private static double Map(double x, double a, double b, double c, double d) {
+      return (x - a) * (d - c) / (b - a) + c;
+    }
+
+    // Map value x from range a-b to range c-d, clamp values outside of range c-d to c or d
+    private static double MapClamp(double x, double a, double b, double c, double d) {
+      return Clamp(Map(x, a, b, c, d), c, d));
```
Comment (Ashoat) [inline Spectrum/Visualizers/LEDDomeRadialVisualizer.cs]: You have an extraneous close paren here, not sure how you even got it to compile?

File: Spectrum/Visualizers/LEDDomeRadialVisualizer.cs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+      }
+    }
+
+    public bool Enabled { get; set; }
+
+    public Input[] GetInputs() {
+      return new Input[] { this.audio };
+    }
+
+    public void Render() {
```
Comment (Ashoat) [inline Spectrum/Visualizers/LEDDomeRadialVisualizer.cs]: Nit: this doesn't seem to need to be `public`

File: Spectrum/Visualizers/LEDDomeRadialVisualizer.cs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+      double progress = this.config.beatBroadcaster.ProgressThroughBeat(1.0);
```
Comment (Ashoat) [inline Spectrum/Visualizers/LEDDomeRadialVisualizer.cs]: Nit: this is just `ProgressThroughMeasure`

**Review by Ashoat — COMMENTED** [review]
(See comments inline.)

These 3 new visualizers are awesome. Thanks so much for all your work this year, I'm super excited to see how it looks on playa!

Regarding `spectrum_default_config.xml`, can you either just add your new config params to it, or alternately delete the `spectrum_config.xml` in the build folder (spectrum/Spectrum/bin/Debug) and run Spectrum to see if missing the new config params in the XML will cause it to crash?

Not for this PR but maybe for a follow-up:

Regarding your question over chat about how to make the visual effects look better, I think your instinct about using the level was right. In my experience the level is  the most live-feeling animation parameter we have to play with. That said, to me the way it's being surfaced right now doesn't feel as visible as it could be. An idea I had would be to vary the length of the radar/spiral based on the level, probably from bottom to top (so the top only lights up when it's loud).

It also would be great to use some more of the colors in the color palette. If for instead the top part had a different color, and only lit up when the level was up near 95% or so, there would be a pretty cool effect on ever downbeat (or whenever it's loudest).

**Comment by budadcabrion** [comment]
I fixed the nits, the extra paren (typo right before committing), added default values to the config, and also tested the default config just to be thorough.

**Pull request merged**